mirror_base.h

```diff
@@ -231,8 +240,101 @@ namespace mirror
 		return nullptr;
 	}
 
+	/*
+     * Some templates to do reflection for functions
+     */
+	template<typename NotAFunction>
+	struct FunctionTraits {}; /// #1
+
+	template<typename R, typename... Args>
+	struct FunctionTraits<R(*)(Args...)> {
+		using result = R;
+		using args = std::tuple<Args...>;
+	};
+
+	template<typename F>
+	using FunctionArguments_T = typename FunctionTraits<F>::args;
+
+	template <std::size_t N, typename F>
+	using FunctionArgument_T = typename std::tuple_element<N, FunctionArguments_T<F>>::type;
+
+	template <typename F, typename T, T I, T Top>
+	struct FunctionArgumentsUnpiler
+	{
+		static void Unpile(class StaticFunction* _function)
+		{
+			using ArgumentType = mirror::FunctionArgument_T<I, F>;
+			_function->addArgument<ArgumentType>();
+			FunctionArgumentsUnpiler<F, T, I + 1u, Top>::Unpile(_function);
+		}
+	};
+
+	template <typename F, typename T, T I>
+	struct FunctionArgumentsUnpiler<F, T, I, I>
+	{
+		static void Unpile(class StaticFunction* _function) {}
+	};
+
+	class StaticFunction : public TypeDesc
+	{
+	public:
+		StaticFunction(size_t _typeHash)
+			: TypeDesc(Type_StaticFunction, "StaticFunction", _typeHash)
+		{
+		}
+
+		template <typename T>
+		void setReturnType()
+		{
+			m_returnType = TypeDescGetter<T>::Get();
+		}
+
+		template <typename T>
+		void addArgument() {
+			TypeDesc* typeDesc = TypeDescGetter<T>::Get();
```

Comment:
Je fais suite à un autre commentaire, ici, pourquoi c'est pas un TypeDesc<T>* qu'on choppe ?

Reply:
De là ou j'en suis dans ma reflexion, le problème des types templatés c'est qu'ils faut connaitre leur type réél pour pouvoir les manipuler en dehors de la compilation, hors le propos de cette lib c'est justement de pouvoir manipuler et déduire des types inconnus à priori au runtime. Ça me parait un peu l'inverse de ce qu'on veut faire du coup...

Reply:
C'est pas l'objectif de mirror à la base je suis d'accord, mais à quoi bon réfléchir les méthodes si tu peux pas les invoquer ? 
Et pour les invoquer, comment faire si on peut pas faire de pont entre TypeDesc et type primitif ? 

J'ai trouvé un truc sympa dans une conf CPPCON de 2016: https://github.com/CppCon/CppCon2016/blob/master/Presentations/C%2B%2B14%20Reflections%20Without%20Macros%2C%20Markup%20nor%20External%20Tooling/C%2B%2B14%20Reflections%20Without%20Macros%2C%20Markup%20nor%20External%20Tooling%20-%20Antony%20Polukhin%20-%20CppCon%202016.pdf

page 31 ça montre une macro sympa qui te permets de faire le pont entre un type et un id et vice-versa.

La conf est super (même si j'en capte que 5%) et j'ai l'impresson que ce qu'il fait pourrait t'intéresser: https://www.youtube.com/watch?v=abdeAew3gmQ

Reply:
Je pense pas que ce soit pas l'objectif de mirror, au contraire t'as raison, invoquer une fonction en pur data driven, ce serait vraiment l'aboutissement du truc. Je pense que par contre c'est pas l'objectif de mirror de te filer des nouveaux outils au moment de la compilation. Pour ça y a déjà les templates qui font bien mieux.

Reply:
Je vais regarder la conf quand j'aurai un moment, ça a l'air chouette

Reply:
Bon, c'était intéressant, mais j'ai vraiment l'impression de regarder des gens essayer de planter un clou avec la langue.
En soi, même sans parler de l'idée de réflexion, ce qu'il fait c'est pas si compliqué conceptuellement (enfin, jusqu'à l'encodage des structs, là j'ai laché), mais c'est tellement l'enfer à écrire et ça vient avec tellement de contraintes que pour moi la conclusion c'est juste que c'est pas le bon outil et qu'il serait temps de passer à autre chose plutôt que d'essayer de réformer ce truc.
Mais bon, dans la vraie vie on a toujours que ça donc pourquoi pas.

Reply:
Si le travail de "reflexion TS" abouti, ça sera tip-top pour faire de la reflexion facile: https://github.com/cplusplus/reflection-ts

Quand tu dis "pas le bon outil", tu parles du language c++ ou de mirror ?

Reply:
Je parle de C++. Je trouve constamment tous ces trucs à base de templates tellement circonvolus par rapport à l'idée de base qu'ils modélisent.

Reply:
j'ai fait quelques recherches, et je pense bien que appeller une fonction par pure réflexion c'est possible.
Avec ça tu peux apparemment appeller une fonction avec en paramètre juste un tuple qui est défini par les bons types : https://stackoverflow.com/questions/7858817/unpacking-a-tuple-to-call-a-matching-function-pointer
Et avec les FunctionTraits que t'as déjà ajouté dans mirror, tu peux avoir le type de ce tuple, donc déclarer une variable tuple qui est déja typée correctement.
Il reste plus qu'a relier les points et remplir le tuple au runtime en fonction des données de mirror (ça pas sur que ce soit possible en vrai, mais pas sur que ce soit impossible non plus)

Reply:
> 
> 
> Je parle de C++. Je trouve constamment tous ces trucs à base de templates tellement circonvolus par rapport à l'idée de base qu'ils modélisent.

Je suis 100% d'accord, en tant q'utilisateur, on ne devrait pas avoir à passer par toute cette complexité. Toutefois, que cette complexité existe ne me choque pas, si elle est cachée derrière des fonction de haut niveau. Imaginons par exemple pour activer la réflection:

```
class A: public std::enable_reflection<A>
{
//
}
```

Et des trucs du genre: `std::invoke_reflected_method( instance, "methodName", param1, param2);`

Reply:
> 
> 
> j'ai fait quelques recherches, et je pense bien que appeller une fonction par pure réflexion c'est possible.
> Avec ça tu peux apparemment appeller une fonction avec en paramètre juste un tuple qui est défini par les bons types : https://stackoverflow.com/questions/7858817/unpacking-a-tuple-to-call-a-matching-function-pointer
> Et avec les FunctionTraits que t'as déjà ajouté dans mirror, tu peux avoir le type de ce tuple, donc déclarer une variable tuple qui est déja typée correctement.

C'est excellent, j'avais pas pensé à ça !

> Il reste plus qu'a relier les points et remplir le tuple au runtime en fonction des données de mirror (ça pas sur que ce soit possible en vrai, mais pas sur que ce soit impossible non plus)

C'est possible car tu peux interroger le type de chaque élément du tuple ! Et cond faire du explicit cast avec.
Sinon si les données que tu veux stocker danss le tuple ont des opérateurs de conversion implicites, ce sera même encore plus simple !

Reply:
Dans Nodable j'ai par exemple une classe `Member` qui peut se caster en `double` `std::string` ou `bool` de manière implicite.

Reply:
ha HA, je viens de réussir à faire marcher un truc assez cool, imaginons :
```
struct MyData
{
	int64_t a;
	float b;
	char* c;

	MIRROR_CLASS_NOVIRTUAL(MyData)
	(
		MIRROR_MEMBER(a)()
		MIRROR_MEMBER(b)()
		MIRROR_MEMBER(c)()
	)
};

bool MyStaticFunction(int _a, float _b, char* _c)
{
	printf("%d,%f,%s\n", _a, _b, _c);
	return true;
}
```

Hé ben je peux faire un truc comme ça :
```
MyData data;
data.a = 2;
data.b = 4.f;
data.c = "lol";

bool ret = CallFunctionWithClassMembersAsArguments(&MyStaticFunction, &data, std::make_tuple( "a", "b", "c" ));
printf("%d\n", ret);
```
et que ça compile, et printe effectivement :
```
2,4.000000,lol
1
```
tout en balancant un assert statique si le nombre d'arguments correspond pas, et des asserts runtime si les types des membres et ceux des arguments correspondent pas.

Reply:
Ça veut dire que dés le moment ou t'as une variable typée vers ton pointeur de fonction, tu peux en pur data driven l'appeller avec n'importe quoi, et je pense que c'est vraiment stylé.

Je me demande quelle est la bonne manière d'interfacer ça dans la librairie, j'avoue que si toi t'avais des cas d'usage concrets ça m'aiderait bien.

Reply:
Oh ! bien sympa tout ça. Je vais essayer de brancher mon submodule de mirror sur ta branche pour te proposer un case d'usage dès que je peux.